### PR TITLE
Revert "revert unpack changes"

### PR DESCRIPTION
**Pull request by aaronbuchwald**
This reverts commit 900eb39da6ae7690c05e64a43589568303cbd403 and migrates the changes from https://github.com/ethereum/go-ethereum/pull/26568.

This was reverted to separate it from the rest of the coreth sync changes.

### CODE REVIEW

**Review by darioush — APPROVED** [review]

**Review by anusha-ctrl — APPROVED** [review]
lgtm seems consistent with geth.

File: accounts/abi/unpack.go [hunk 1]
```diff
@@ -43,43 +44,72 @@ var (
...
+			}
+			return uint32(u64), nil
 		case 64:
-			return binary.BigEndian.Uint64(b[len(b)-8:])
+			if !isu64 {
+				return nil, errBadUint64
+			}
+			return u64, nil
 		default:
 			// the only case left for unsigned integer is uint256.
```
Comment (morrisettjohn) [inline accounts/abi/unpack.go:74]: Could we make this a case 256 and make default return an error?  I think it's better to be overly cautious, and if something incredibly bizarre does happen where typ.size is an invalid length, it's going to be a lot easier to find the issue.

File: accounts/abi/unpack.go [hunk 2]
```diff
@@ -43,43 +44,72 @@ var (
...
+		}
+		return int32(i64), nil
 	case 64:
-		return int64(binary.BigEndian.Uint64(b[len(b)-8:]))
+		if !isi64 {
+			return nil, errBadInt64
+		}
+		return i64, nil
 	default:
 		// the only case left for integer is int256
```
Comment (morrisettjohn) [inline accounts/abi/unpack.go:110]: Same comment as above

**Review by morrisettjohn — CHANGES_REQUESTED** [review]
LGTM, I have two nits that should never be an issue - but then again you never know!

File: accounts/abi/unpack.go (hunk 1 above)
Comment (aaronbuchwald) [reply accounts/abi/unpack.go:74]: I don't think there's a need to be overly cautious here. There's a finite set of valid int/uint types and I'd prefer to keep this aligned with upstream.

**Pull request merged**